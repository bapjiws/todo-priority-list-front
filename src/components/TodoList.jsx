import React, { Component } from 'react';
import { connect } from 'react-redux';

// TODO: get rid of mockloadTodos when done
import { loadTodos, mockloadTodos } from '../../redux/actions/todos';

export class TodoList extends Component {
    componentDidMount() {
        this.props.mockloadTodos();
    }

    render() {
        return <table className="todos-container">
            <thead>
                <tr>
                    <th>Name</th>
                    <th>Description</th>
                    <th>Priority</th>
                </tr>
            </thead>
            <tbody>
                {
                    this.props.todos.map((todo, idx) => <tr key={idx}>
                        <td>{todo.name}</td>
                        <td>{todo.description}</td>
                        <td>{todo.priority}</td>
                    </tr>)
                }
            </tbody>
        </table>
    }
}

export default connect(
    state => ({
        todos: state.todos.data
    }),
    { loadTodos, mockloadTodos }
)(TodoList);