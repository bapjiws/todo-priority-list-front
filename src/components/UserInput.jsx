import React, { Component } from 'react';
import { connect } from 'react-redux';
import { FormGroup, FormControl, ControlLabel, Radio, Button } from 'react-bootstrap';

import { addTodo } from '../../redux/actions/todos';

const priorityScale = [1,2,3,4,5,6,7,8,9,10];

export class UserInput extends Component {
    constructor(props) {
        super(props);
        this.state = {
            priority: 0,
            name: '',
            description: ''
        };

        // this.priorities = priorityScale.reduce((acc, curr) => {
        //     acc[curr] = null;
        //     return acc;
        // }, {});

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleCLick = this.handleCLick.bind(this);
    }

    handleChange(event) {
        // console.log('event:', event.target.id, event.target.value);

        const { id, value } = event.target;
        this.setState({
            [id]: id === 'priority' ? +value : value // this.state.priority is now handled by this.handleCLick
        });
    }

    handleSubmit(event) {
        // console.log('this.getNameValidationState():', this.getNameValidationState());

        if (this.getNameValidationState() === 'error' || this.getDescriptionValidationState() === 'error') {
            console.log('INPUT ERROR');
            event.preventDefault();
            return;
        }

        console.log('BUTTON');

        event.preventDefault();

        this.props.addTodo(this.state);
        document.querySelectorAll('input[type="radio"]')[this.state.priority-1].checked = false;
        this.setState({
            priority: 0,
            name: '',
            description: ''
        });
    }

    handleCLick(event) {
        // console.log('event:', event.target);
        console.log('CLICK:', +event.target.value);
        this.setState({
           priority: +event.target.value
        });
    }

    getNameValidationState() {
        const length = this.state.name.length;
        return length === 0 ? 'error' : 'success';
    }

    getDescriptionValidationState() {
        const length = this.state.description.length;
        return length === 0 ? 'error' : 'success';
    }

    render() {
        return <form className="form-container">
            <FormGroup
                className="form-space-between"
                controlId="priority"
            >
                <ControlLabel>Priority</ControlLabel>
                {' '}
                <span>
                    {
                        priorityScale.map((priority, idx) =>
                            <Radio
                                inline key={idx}
                                name="radioGroup"
                                value={priority}
                                // inputRef={ref => { this.priorities[priority] = ref; }}
                                onClick={this.handleCLick}
                            >
                                {priority}
                            </Radio>)
                    }
                </span>
            </FormGroup>

            <FormGroup
                controlId="name"
                validationState={this.getNameValidationState()}
            >
                <ControlLabel>Name</ControlLabel>
                <FormControl
                    type="text"
                    value={this.state.name}
                    placeholder="Name your todo"
                    onChange={this.handleChange}
                />
                <FormControl.Feedback />
            </FormGroup>

            <FormGroup
                controlId="description"
                validationState={this.getDescriptionValidationState()}
            >
                <ControlLabel>Description</ControlLabel>
                <FormControl
                    type="text"
                    value={this.state.description}
                    placeholder="Describe your todo"
                    onChange={this.handleChange}
                />
                <FormControl.Feedback />
            </FormGroup>

            <Button type="submit" onClick={this.handleSubmit}>
                Submit
            </Button>
        </form>
    }
}

export default connect(
    null,
    { addTodo }
)(UserInput);