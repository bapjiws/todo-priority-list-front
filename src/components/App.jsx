import React from 'react';
import { Provider } from 'react-redux';

import configureStore from '../../redux/configureStore';

import '../../styles/main';

import UserInput from './UserInput';
import TodoList from './TodoList';

const App = () => {
    return <Provider store={configureStore()}>
        <div>
            <UserInput/>
            <TodoList/>
        </div>
    </Provider>
};

export default App;