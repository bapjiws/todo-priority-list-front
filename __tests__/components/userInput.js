import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json'; // Documentation: https://github.com/adriantoine/enzyme-to-json

import { UserInput } from '../../src/components/UserInput';

describe('UserInput', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<UserInput />)
    });

    it('renders correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should update pieces of its own state', () => {
        wrapper.find('input[name="priority"]').simulate('change', { target: { name: 'priority', value: 3 } });
        expect(wrapper.state('priority')).toBe(3);

        wrapper.find('input[name="name"]').simulate('change', { target: { name: 'name', value: 'do laundry' } });
        expect(wrapper.state('name')).toBe('do laundry');

        wrapper.find('input[name="description"]').simulate('change', { target: { name: 'description', value: 'quickly!' } });
        expect(wrapper.state('description')).toBe('quickly!');
    })
});