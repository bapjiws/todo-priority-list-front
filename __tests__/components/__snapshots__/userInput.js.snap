// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInput renders correctly 1`] = `
<form
  className="form-container"
>
  <span
    className="input-row"
  >
    <span>
      Priority:
    </span>
    <input
      name="priority"
      onChange={[Function]}
      type="number"
      value={0}
    />
  </span>
  <span
    className="input-row"
  >
    <span>
      Name:
    </span>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
  </span>
  <span
    className="input-row"
  >
    <span>
      Description:
    </span>
    <input
      name="description"
      onChange={[Function]}
      type="text"
      value=""
    />
  </span>
  <button
    onClick={[Function]}
  >
    Add
  </button>
</form>
`;
